engine/consensus/approvals/assignment_collector.go

```diff
@@ -0,0 +1,346 @@
+package approvals
+
+import (
+	"fmt"
+	"math/rand"
+	"sync"
+
+	"github.com/rs/zerolog/log"
+
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/engine/consensus/sealing"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/flow/filter"
+	"github.com/onflow/flow-go/model/messages"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/mempool"
+	"github.com/onflow/flow-go/network"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/storage"
+)
+
+// helper functor that can be used to retrieve cached block height
+type GetCachedBlockHeight = func(blockID flow.Identifier) (uint64, error)
+
+// AssignmentCollector is responsible collecting approvals that satisfy one assignment, meaning that we will
+// have multiple collectorTree for one execution result as same result can be incorporated in multiple forks.
+// AssignmentCollector has a strict ordering of processing, before processing approvals at least one incorporated result has to be
+// processed.
+// AssignmentCollector takes advantage of internal caching to speed up processing approvals for different assignments
+// AssignmentCollector is responsible for validating approvals on result-level(checking signature, identity).
+// TODO: currently AssignmentCollector doesn't cleanup collectorTree when blocks that incorporate results get orphaned
+// For BFT milestone we need to ensure that this cleanup is properly implemented and all orphan collectorTree are pruned by height
+// when fork gets orphaned
+type AssignmentCollector struct {
+	ResultID                             flow.Identifier                        // ID of execution result
+	result                               *flow.ExecutionResult                  // execution result that we are collecting approvals for
+	BlockHeight                          uint64                                 // height of block targeted by execution result
+	collectors                           map[flow.Identifier]*ApprovalCollector // collectors is a mapping IncorporatedBlockID -> ApprovalCollector
+	authorizedApprovers                  map[flow.Identifier]*flow.Identity     // map of approvers pre-selected at block that is being sealed
+	lock                                 sync.RWMutex                           // lock for protecting collectors map
+	verifiedApprovalsCache               *ApprovalsCache                        // in-memory cache of approvals (already verified)
```

Comment:
I feel the cache has unnecessary high complexity here:
* I don't think we need an LRU cache, as we know the max capacity should not be exceeded. (we could panic otherwise or just reject new elements)
* do we need a cache here, that tracks approvals by resultID? 

I feel a map from key: `approval.Body.Attestation.ID()` to `ResultApproval` should be sufficient. We should document why it can't overflow: 
> An assignment collector is dedicated to a specific pair of `BlockID` and `ExecutionResultID`. For a given `BlockID` and `ExecutionResultID` each verifier can only provide one approval per `ChunkIndex`, which is guaranteed by `approval.Body.Attestation.ID()`. Hence, byzantine verifiers can't overflow this cache.

Reply:
and what is exactly meant by overflowing? We will need to track how many elements in map?